logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,158 @@
+                if (isSelfDeletionOutgoing) return@launch
+
+                ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                    currentState + ((conversationId to messageId) to SelfDeletionTimeLeft(0))
+                }
+            }
+
+            messageRepository.getMessageById(conversationId, messageId).map { message ->
+                require(message is Message.Ephemeral)
+
+                enqueueMessageDeletion(message)
+            }
+        }
+    }
+
+    private fun enqueueMessageDeletion(message: Message.Ephemeral) {
+        userSessionCoroutineScope.launch {
+            val selfDeletingMessageTimer = SelfDeletingMessageTimer(
+                coroutineScope = userSessionCoroutineScope
+            )
+
+            for (timerEvent in selfDeletingMessageTimer.startTimer(message.timeLeftForDeletion())) {
+                when (timerEvent) {
+                    is SelfDeletionTimerState.Started -> {
+                        if (!message.isDeletionScheduledInThePast()) {
+                            messageRepository.markSelfDeletionStartDate(
+                                message.conversationId,
+                                message.id,
+                                Clock.System.now().toEpochMilliseconds()
+                            )
+                        }
+                    }
+
+                    is SelfDeletionTimerState.OnGoing -> {
+                        println("time left ${timerEvent.timeLeft.timeLeftInMs}")
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState + ((message.conversationId to message.id) to timerEvent.timeLeft)
+                            }
+                        }
+                    }
+
+                    is SelfDeletionTimerState.Finished -> {
+                        messageRepository.deleteMessage(message.id, message.conversationId)
+                        ephemeralMessageDeletionTimeLeftMapMutex.withLock {
+                            ephemeralMessageDeletionTimeLeftMap.update { currentState ->
+                                currentState - (message.conversationId to message.id)
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+    }
+
+    override fun observePendingMessageDeletionState() = ephemeralMessageDeletionTimeLeftMap
+
+    override fun enqueuePendingSelfDeletionMessages() {
+        launch {
+            messageRepository.getEphemeralMessages()
+                .onSuccess { ephemeralMessages ->
+                    ephemeralMessages.forEach { ephemeralMessage ->
+                        require(ephemeralMessage is Message.Ephemeral)
+
+                        enqueueMessageDeletion(message = ephemeralMessage)
+                    }
+                }
+        }
+    }
```

Comment:
Does this means that it doesn't wait for the message to be read before it starts deleting stuff?

Reply:
Good point I need the ephemeral messages here that are already been marked as started then yes we enqueue those because they were already read and started